feat!: upgrade support to node 22 and 24

kmruiz opened the pull request:
We are also upgrading the minimum engine to 22, and the dependencies we use for CodeQL to 24, as it's going to be our baseline from now on.

kmruiz committed:
chore: Upgrade support to node 22 and 24

We are also upgrading the minimum engine to 22,
and the dependencies we use for CodeQL to 24, as it's
going to be our baseline from now on.

kmruiz committed:
chore: use devtools toolchain on linux and upgrade macos deps

kmruiz committed:
chore: change display name, it's misleading

kmruiz committed:
chore: add brotli as a dependency when compiling the test

kmruiz committed:
chore: reuse devtools toolchain in Linux

kmruiz committed:
chore: fix typo

kmruiz committed:
chore: simplify by using a stable folder after installation

kmruiz committed:
chore: Remove unsupported flags for r/o heaps

According to a v8 commit, v8 removed the ability
to build without shared r/o heaps.

More info: https://chromium-review.googlesource.com/c/v8/v8/+/6038531

kmruiz committed:
chore: fix check in OSX

kmruiz committed:
chore: use 22 until we have 24

kmruiz committed:
chore: create parent node folder, necessary for the symlink

kmruiz committed:
chore: upgrade node-gyp to the latest

kmruiz committed:
chore: upgrade test template to node 24 and std=c++20

kmruiz committed:
chore: remove assert(CanUseCustomSnapshotPerIsolate)

We only use one EmbedderSnapshotData and one Isolate, so we are
not hitting the corner case of having isolates with different
snapshots.

kmruiz committed:
chore: forgot to remove the test skip

addaleax committed:
chore: unregister isolate *after* disposing isolate

Refs: https://github.com/nodejs/node/commit/9aa1afb52741498345ec483ff72932520c5050ff
Refs: https://github.com/nodejs/node/commit/5d3e1b555c0902db1e99577a3429cffedcf3bbdc

addaleax committed:
feat!: remove support code for older Node.js versions

addaleax committed:
chore: add evergreen task for testing dchecks

addaleax committed:
chore: use ubuntu instead of amazon for dcheck runs

addaleax committed:
fixup: use fixed version of weak-napi

kmruiz committed:
chore: upgrade to vs2022 to test

kmruiz committed:
chore: upgrade host so it has msvc 2022

kmruiz committed:
chore: typo in the evergreen distro

addaleax committed:
chore(ci): use llvm from homebrew on evergreen macOS

kmruiz committed:
chore: Use python3.11 in windows host

kmruiz committed:
chore: make sure python 3.11 is used everywhere, also when compiling

kmruiz committed:
chore: make sure that the hash is updated so it uses the proper python version

kmruiz committed:
chore: also setup python and python3 env vars for node-gyp

kmruiz committed:
chore: use NODE_GYP_FORCE_PYTHON

kmruiz committed:
chore: use windows path and simplify

kmruiz commented on .evergreen/use-node.sh:
```diff
@@ -1,8 +1,15 @@
+    export PATH="/opt/devtools/bin:$PATH"
+
+    if [ "$(uname -s)" == "Darwin" ] ; then # in OSX use nvm
+	export NVM_DIR="$PWD/.deps/nvm"
```
ah true, forgot to setup spaces for my indent instead of tabs, I'll use spaces to be consistent with what we had.

kmruiz committed:
chore: normalise to spaces

addaleax commented on .evergreen/install-node.sh:
```diff
@@ -21,12 +24,18 @@ if [[ "$OS" == "Windows_NT" ]]; then
-  [ -s "${NVM_DIR}/nvm.sh" ] && source "${NVM_DIR}/nvm.sh"
-  nvm install --no-progress "$NODE_VERSION"
+    if [ "$(uname -s)" == "Darwin" ] ; then # install Node.js on MacOS
+	curl -o- $NVM_URL | bash
```
@kmruiz still seeing tabs here :)

kmruiz committed:
chore: untabify, use spaces always

kmruiz committed:
chore: setup VC environment and upgrade command args for vcbuild

kmruiz committed:
chore: stub APPDATA env var

kmruiz committed:
chore: adapt it to use CMD properly

kmruiz committed:
chore: fix order of parameters

kmruiz committed:
chore: we still need to projgen

kmruiz committed:
chore: do not use $TEMP because files there can be deleted

kmruiz committed:
chore: typo :(

kmruiz committed:
chore: upgrade to node 24 (the LTS) and increase timeout

The build in Windows is slower, so we increase the timeout
to avoid issues related to slow compilation

kmruiz committed:
chore: use the same base in all machines (22.17.0)

kmruiz committed:
chore: setup windows tests also in GHA for faster feedback

This also adds a baseline: if the tests without special setup
work here, it should work also in evergreen hosts if properly
configured.

kmruiz committed:
chore: remove unused deps and remove setup of MSVC DevTools

In theory vcbuild.bat should detect where the MSVC DevTools
are located (using vswhere.exe) and setup them correctly. We
are going to verify this behaviour.

kmruiz committed:
chore: support all architecture flags

kmruiz committed:
chore: increase timeout (default is 6h, builds can take more time)

kmruiz committed:
chore: fix linter issues

kmruiz committed:
chore: use DEBUG, as it's faster, and it's good enough for a safety check

kmruiz committed:
chore: split each test in it's own job to avoid disk space issues

kmruiz committed:
chore: we can't use cache: npm because we don't have a package-lock

kmruiz committed:
chore: delete precompiled headers on windows and node24>

Compiling with a snapshot requires compiling twice: a base image
and the image with the snapshot embedded. In that situation, clang-cl
complains that there are precompiled headers that changed between
compilations and does not refresh them, but kills the compilation
process with an error. Due to this, before attempting the second
compilation, we will delete all pch files.

kmruiz committed:
chore: Remove test, now the behaviour is tested by the compiler

This test was written so we could speed up the feedback loop.

kmruiz committed:
chore: use rimraf instead of custom code, it should be easier to maintain

kmruiz commented on test/index.ts:
```diff
@@ -6,13 +6,25 @@ import childProcess from 'child_process';
+
+    assert(JSON.stringify(deletedFiles) === JSON.stringify(['delete_1.pch', 'inner/delete_2.pch']));
+  });
+});
```
What I'm thinking is replacing the function with rimraf and a glob filter (it's already a dep in the project), that should work and don't require a test. I'll try to simplify this while I have some tests running in the background.

kmruiz committed:
chore: Generate the GHA workflow from a template

Because we can keep adding more tests, maintaining the GHA workflow
can become a bit cumbersome. This script generates the GHA file
from a template, and adds all the test cases in a matrix. We do this
because each test is expensive in time and disk space, so we do
sometimes hit the GHA limit of 6h or the we fill the disk.

kmruiz committed:
chore: Fix linter issues (spacing and unused deps)

kmruiz committed:
chore: use -f instead of -g for the test runner

It seems that -g expects a regular expression, and
if the regular expression is invalid, mocha just fails
with Error: null. Using -f, uses a substring comparison,
which is what we need.

kmruiz committed:
chore: Remove pch files from compilation artifacts in tmp directory

kmruiz committed:
chore: Print the folder where we are deleting pchs from

kmruiz committed:
chore: attempt to upgrade rimraf to latest version

kmruiz committed:
chore: Use follow: true, so it follows symlinks and junctions

kmruiz committed:
chore: resolve the realpath of nodeSourcePath before deleting pchs

nodeSourcePath in Windows refers to a user directory that contains
an alias. For example, something like C:\Users\RUNNER~1 (in GHA) while
MSVC uses the full name like C:\Users\runneradmin. This can make
rimraf not delete the pch files properly. By resolving first, we make
sure that both paths are absolute and unambiguous.

kmruiz committed:
chore: Remove all directories that can contain compilation outputs

clang-cl does not properly handle pch refresh when a source header
changed after the pch file was generated

kmruiz commented on src/index.ts:
```diff
@@ -350,25 +352,11 @@
-    };
+  const { customCodeSource, customCodeConfigureParam, customCodeEntryPoint } = {
+    customCodeSource: path.join(nodeSourcePath, 'lib-boxednode', `${namespace}.js`),
+    customCodeConfigureParam: `./lib-boxednode/${namespace}.js`,
```
It is in fact user input, and it's used for the compilation output.

It uses the namespace configuration options to provide a namespaced _require_ in the final executable.

kmruiz replied:
It is in fact user input, and it's used for the compilation output.

It uses the namespace configuration options to provide a namespaced _require_ in the final executable.

kmruiz committed:
chore: try to delete the whole out directory for a fresh compilation

kmruiz committed:
chore: Simplify naming, we now wipe out the whole out directory

This will force Node.js to compile from scratch. It's slower, but
it will only happens on Windows and it will be safer and more
reliable there.

kmruiz committed:
chore: ensure that any output directory is cleaned up before compiling

Right now it's failing on the first compilation of the second test, as
we are not doing any clean up. To ensure that we are in a consistent
state in Windows, before attempting compiling, we will make sure we
don't have any old compilation artifact.

This might slow down Windows compilations as they can't reuse cached
files.
